src/pyroboplan/planning/prm.py

```diff
@@ -0,0 +1,287 @@
+            if time.time() - t_start > timeout:
+                print(f"Roadmap construction timed out after {timeout} seconds.")
+                break
+
+            # At each iteration we naively sample a valid random state and attempt to connect it to the roadmap.
+            q_sample = next(sample_generator)
+            if check_collisions_at_state(self.model, self.collision_model, q_sample):
+                continue
+
+            # It's a valid configuration so add it to the roadmap
+            new_node = Node(q_sample)
+            self.graph.add_node(new_node)
+            self.connect_node(new_node, self.options.max_neighbor_radius)
+            added_nodes += 1
+
+    def connect_node(self, new_node, radius):
+        """
+        Identifies all neighbors and makes connections to the added node.
+
+        Parameters
+        ----------
+            parent_node : `pyroboplan.planning.graph.Node`
+                The node to add.
+
+        Returns
+        -------
+            bool :
+                True if the node was connected to the graph. False otherwise.
+        """
+        # Add the node and find all neighbors.
+        neighbors = self.graph.get_nearest_neighbors(new_node.q, radius)
+
+        # Attempt to connect at most `max_neighbor_connections` neighbors.
+        ret = False
+        for node, _ in neighbors[0 : self.options.max_neighbor_connections]:
+            # If the nodes are connectable then add an edge.
+            if has_collision_free_path(
+                node.q,
+                new_node.q,
+                self.options.max_angle_step,
+                self.model,
+                self.collision_model,
+            ):
+                self.graph.add_edge(node, new_node)
+                ret |= True
+
+        return ret
+
+    def reset(self):
+        """
+        Resets the PRM's transient data between queries.
+        """
+        self.latest_path = None
+        for node in self.graph.nodes:
+            node.parent = None
+            node.cost = None
+
+    def plan(self, q_start, q_goal):
+        """
+        Plans a path from a start to a goal configuration using the constructed graph.
+
+        Parameters
+        ----------
+            q_start : array-like
+                The starting robot configuration.
+            q_goal : array-like
+                The goal robot configuration.
+        """
+
+        # Check start and end pose collisions.
+        if check_collisions_at_state(self.model, self.collision_model, q_start):
+            print("Start configuration in collision.")
+            return None
+        if check_collisions_at_state(self.model, self.collision_model, q_goal):
+            print("Goal configuration in collision.")
+            return None
+
+        # Ensure the start and goal nodes are in the graph.
+        start_node = Node(q_start)
+        self.graph.add_node(start_node)
+        goal_node = Node(q_goal)
+        self.graph.add_node(goal_node)
```

Comment:
One other comment that I just realized.

I think at the end of planning you actually want to remove `start_node` and `goal_node` from the graph.

This means that the graph class will also need a `remove_node` method that removes all edges connected to said node.

Reply:
pyrobosim's graph has this already btw, feel free to copypasta unless the underlying data structure changes make that useless

Reply:
I added and found some issues with the fact that we can add "duplicate" edges between nodes, so I made it so that the "least" node is always stored as "nodeA". Sorry to keep blowing this up.

I think it makes more sense to hide that from the caller anyway, so now `add_edge` and `remove_edge` have the same interface.